src/cmap/connection_pool.ts

```diff
@@ -431,19 +437,51 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
       );
       return;
     }
-
     // handle non load-balanced case
+    const interruptInUseConnections = options.interruptInUseConnections ?? false;
+    const oldGeneration = this[kGeneration];
     this[kGeneration] += 1;
     const alreadyPaused = this[kPoolState] === PoolState.paused;
     this[kPoolState] = PoolState.paused;
 
     this.clearMinPoolSizeTimer();
     if (!alreadyPaused) {
-      this.emit(ConnectionPool.CONNECTION_POOL_CLEARED, new ConnectionPoolClearedEvent(this));
+      this.emit(
+        ConnectionPool.CONNECTION_POOL_CLEARED,
+        new ConnectionPoolClearedEvent(this, { interruptInUseConnections })
+      );
     }
+
+    if (interruptInUseConnections) {
+      process.nextTick(() => this.interruptInUseConnections(oldGeneration));
+    }
+
     this.processWaitQueue();
   }
 
+  /**
+   * Closes all checked in perished connections in the pool with a resumable PoolClearedOnNetworkError.
+   *
+   * If interruptInUseConnections is `true`, this method attempts to kill checked out connections as well.
+   * Only connections where `connection.generation <= minGeneration` are killed.  Connections are closed with a
+   * resumable PoolClearedOnNetworkTimeoutError.
+   */
+  private interruptInUseConnections(minGeneration: number) {
+    for (const connection of this[kCheckedOut]) {
+      if (connection.generation <= minGeneration) {
+        this[kCheckedOut].delete(connection);
+        connection.onError(new PoolClearedOnNetworkError(this));
+        this.emit(
```

Comment:
I think for the sake of consistency, we *should* use the `this.destroyConnection` method instead of just "faking" the event emission here; if we just do the "onError" call on the connection, it won't properly fall through to the connection's own destroy method which is what's responsible for clearing the connection's event listeners, and even though right now those event listeners are only for lb mode, it's conceivable that we might have other listeners in the future and we don't want to accidentally introduce a leak. The connection's destroy method outside the listener removal is a no-op on an already destroyed stream, so this should be safe; however, we should also update the connection's `onError` method to also set `this.destroyed = true`. Then, with this refactor, in the `checkIn` method, we should be able to check for `connection.destroyed` directly, and instead of the misleading `willDestroy` that is only a `willDestroyIfNotDeleted`, we can update all that as follows:
- rename `willDestroy -> shouldDestroy`(or `isUsable` or `shouldReturnToThePool`, but something that communicates that the point is to determine whether we can reuse the connection, just be sure to flip the `!` if flipping the meaning)
- update check in following line to `if (!shouldDestroy)`
- update L399 to not worry whether the connection has already been deleted from the checked out set
- update L402 to `if (shouldDestroy && !connection.destroyed)` (you could also call `connection.destroyed -> isDestroyed`)